### ISSUE TITLE
Why did you set temperature to 0.1?

### ISSUE THREAD

**Issue by callmesora** [lchen001/LLMDrift#2]
Hey! 
Intresting work :) 

Why did you guys set the temperature of the models to such a low number? 
I think GPT4 works with MOE  and they optimized the usage of these experts since march so I reacon it's possible it performs worse with lower temperatures due to these optimizations but perform just as good with higher ones.

Thanks for the study!

**Comment by nixtoshi** [comment]
Hey @callmesora I'm also a curious about these findings because I had a suspicion that this might be the case.

Where did you read the test are being done with 0.10 temperature?

If this is the case, the findings would not be valid because the temperature variable willingly adds chaos to the generation.

Any test that tests for factualness should only use temperature 0 (no chaos, deterministic generations) with the same configuration every time

**Comment by ChatGPT4U** [comment]
Based on the data in the .csv files, it is evident that the temperature was set to 0.1.

Please read my article, "[How Two Papers Treated Those Prime Numbers Differently](https://realpromptguru.gumroad.com/p/how-two-papers-treated-those-prime-numbers-differently)", for more details.

**Comment by nixtoshi** [comment]
> Based on the data in the .csv files, it is evident that the temperature was set to 0.1.

If that's the case, that would invalidate these findings, wouldn't it?

**Comment by elmstedt** [comment]
> > Based on the data in the .csv files, it is evident that the temperature was set to 0.1.
> 
> If that's the case, that would invalidate these findings, wouldn't it?

It doesn't necessarily invalidate them—from a statistical perspective—but it would have been better if they either chose `temperature = 0` or had replicated each prompt at whichever temperature they chose so as to capture the variability.

Though, a potential reason to consider the findings invalid would be because the models are not typically used at that temperature. The default is `temperature = 0.7`, so that almost certainly was the temperature used during initial training and the RLHF process, so the models will almost certainly perform better *overall* at this "native" temperature than they do at a much lower temperature.

**Comment by sovos15** [comment]
just to go over what seems to be agreed on here. 

Higher temperature values add chaos.

 Most users use 0.7
  This test uses 0.1

 Having not used 0.0 invalidates testing 
   Having not used 0.7 invalidates testing

The program would work better at 0.0
 The program would work better at 0.7

  It's kind of fantastic to be having a conversation like this at all when you could just go try the gpt's to see that they can no longer do math. Doesn't look like Bard, Claude, or Bing are doing much better. Personal experience, quality has dropped significantly. So you can sit around here trying to "invalidate" the experiment rather than repeating it. Seeing for yourself that it is repeatable. I mean they went so far as to call it drift. Repeating the experiment over the next 3 months could very well find math improve. So the theory is absolutely sound and If you find yourself in the 0.7 camp then you need to do the test yourself at 0.7. If you find yourself in 0.0 test do the test at 0.0 ffs...

  I personally think that if you folks are so damn uncertain about what's going to work you should do both. In fact you should do every single temperature between 0.0 and 0.9 just to create more data. More data is better. Sitting around talking contradictions is useless.

**Comment by elmstedt** [comment]
> It's kind of fantastic to be having a conversation like this at all when you could just go try the gpt's to see that they can no longer do math. [...] Personal experience, quality gas dropped significantly...

That's exactly what we're seeking to avoid, relying on personal experience and conjecture.

That's the point of doing *science*.

Many of us here are just frustrated and pointing out that this is not good science.

**Comment by sovos15** [comment]
Sitting around pointing out that it's not good science is not good science.
If you see a problem with the experiment you must repeat it with the
modifications to see if it actually makes the change that you think that it
will before you can say that there was any change to be made.

**Comment by elmstedt** [comment]
That's not how science works my friend.

Consider this an unofficial peer-review, since the authors have not submitted their paper to any journals yet—likely because it won't pass peer-review—because of the issues people have already identified here.

**Comment by sovos15** [comment]
It's okay Hitler's mine compf reworded to be woke passes peer review so I'm
not too worried about if or not something can pass a peer review
process.

 I'm worried about whether or not you could pass basic math.

 What would be the right answer? 

Serious question if more and less are both the
wrong answer what is the right answer?

 You've completely shirked your
responsibility as a scientist to do experiments yourself and relied on this
broken backwards peer review concept that doesn't have any place in the
formalized scientific process...

 Do you realize how funny you are?

**Comment by elmstedt** [comment]
Alright, I think we can safely be done with ending this discussion.

Good luck friend.

**Comment by sovos15** [comment]
Long story short "there is no right answer because you disagree with the
findings." You will continue to move the goal posts as long as far as you
need to and having been asked to plant them firmly anywhere you will run
for the hills

**Comment by sovos15** [comment]
Not only do you not know where the goal posts go you don't know which goal is yours. You need serious help...

For posterity, if any sane people come around with questions about just how stable my position is here.

"Should the experiment have been run at a temperature of 0.0 or 0.7 in order to be considered valid?"

This contradiction is plain as day. It's astounding multiple people fell into it prior to having been corrected. 

Specifically.
"What's the right temperature for a valid experiment?" Shouldn't be hard to answer if you're going to invalidate the experiment based on having used to the wrong temperature. They can't even decide if it needs to go up or down...


For anyone with questions about my comment on peer review.

https://en.m.wikipedia.org/wiki/Grievance_studies_affair

Another repeatable experiment...

**Comment by nixtoshi** [comment]
> Not only do you not know where the goal posts go you don't know which goal is yours. You need serious help...
> 
> For posterity, if any sane people come around with questions about just how stable my position is here.
> 
> "Should the experiment have been run at a temperature of 0.0 or 0.7 in order to be considered valid?"
> 
> This contradiction is plain as day. It's astounding multiple people fell into it prior to having been corrected.
> 
> Specifically. "What's the right temperature for a valid experiment?" Shouldn't be hard to answer if you're going to invalidate the experiment based on having used to the wrong temperature. They can't even decide if it needs to go up or down...
> 
> For anyone with questions about my comment on peer review.
> 
> https://en.m.wikipedia.org/wiki/Grievance_studies_affair
> 
> Another repeatable experiment...

Temperature is specifically designed to generate random responses within a range of responses. Since this is a factualness test (prime number test), adding temperature will create random responses and will make the model invent false responses ON PURPOSE.

The fact that this test was done with any amount of temperature invalidates the findings, because they could be attributed to chance.

Your responses don't really answer WHY temperature was used in this paper. I believe nor you nor the authors of this paper understand temperature in LLMs.

**Comment by elmstedt** [comment]
> Temperature is specifically designed to generate random responses within a range of responses. Since this is a factualness test (prime number test), adding temperature will create random responses and will make the model invent false responses ON PURPOSE.

One point to make here. This is incorrect. You're making the assumption that the sequence consisting of most-likely next tokens will always result in the best answer., it does not. Otherwise temperature would *always* be zero.

By way of example, say there are four possible first tokens, with probablities 1/3, 2/9, 2/9, 2/9. Then let's say that the generated sequence following the greedy selection of the first token always fails, while choosing any of the other first tokens leads to a successful sequence.

Hopefully you can see why we wouldn't want to set the temperature to zero in this instance—a model which gets the correct answer 2/3 of the time would appear to always fail.

The primary reason for setting `T = 0` is for reproducibility.

What they probably should have done though is use a temperature in a normal range, then do several replications of each test.

> The fact that this test was done with any amount of temperature invalidates the findings, because they could be attributed to chance.

Again, having randomness does not, in and of itself, invalidate the experiment. But if they were going to include randomness, it should have been done with a more normal temperature with replications.

At the end of the day though the prime evaluation test is a bad test because it's just not something the model is capable of doing, nor is it something the model was ever capable of doing.

**Comment by nixtoshi** [comment]
@elmstedt While it is true that the most likely response isn't always the best response or the closest to the truth, very often it IS, specially for widely known facts found on the internet. This is why the settings for chatbots designed to be as factual as they can be, usually ONLY USE a temperature of 0. For instance, chatGPT uses a temperature of 0.

Using a temperature above 0 does have a lot of uses, mainly in creative generations, entertainment and sounding more human-like. For instance, I've made chatbots that were designed to be fun to chat with, and a temperature 0 doesn't work in this case, because every time, it had the same responses, and the chatbot told the exact same jokes making it less fun to talk to. But with a temperature of 0.40, the responses were varied and still good for this non-critical application, telling jokes.

Since OpenAI trained its model on a lot of common crawl, it is unlikely that most people get a known fact wrong, such as wether X number is a prime, odd or even.

Therefore GPT3 and GPT4 CAN known which numbers are prime and which aren't, specially for commonly queried numbers that are commonly found online, which are probably the first 1000 prime numbers~. Beyond those common primes, sure, it would fail, specially for large prime numbers that it has no prior information of, however, it could identify some common patterns, like the number ending in a even number to deduce that a very large number it hasn't seen before isn't prime.

This ability to correctly give a response without prior knowledge is called zero-shot learning.

**Comment by elmstedt** [comment]
@nixtoshi you seem to have some financial misunderstandings about the technology and how it works, so I think we can be done here.

I stand by what I've written.

**Comment by sovos15** [comment]
Still to this day people can't agree whether it should be absolute zero or
higher then 0.1 so let me add some information beyond "if you don't like
the experiment it's repeatable do it yourself under your own conditions and
see if anyone cares".

 If I understand the original post MOE systems require higher temperature
due to "parallelism" and "slicing" now I'm a not really into technobabble
but as a programmer it sounds like that's aliasing and dirivitives. If the
net results is a lighter network then some dirivitives serve as the target
of multiple aliases and or aliases far outnumber dirivitives and
dirivitives increases quality to make up for the simplification of aliasing.

Someone would have to know what the case is to answer if more or less
temperature is correct without doing an array of tests and regardless of
all that the point of the study remains.

Don't get me started on the hyper focus on just one of the eight or so
tests. You people are broken...


 Tl:Dr Given a group of settings that work at a given time seeing those
same settings decrease in efficacy over time is data. Don't ask the person
to have used different settings to start with make more data you lazy sad
sacks.
